### ISSUE TITLE
electron-compile support

### ISSUE THREAD

**Issue by dignifiedquire** [electron-userland/electron-builder#807]
- **Version**: `7.11.3`

- **Target**: `all`

I've just integrated my app with [`electron-prebuilt-compile`](https://github.com/paulcbetts/electron-prebuilt-compile) which compiles various assets and JavaScript using babel. According to https://github.com/electron/electron-compile#how-can-i-precompile-my-code-for-release-time it works with `electron-packager`. I tried just running the build, but nothing gets build with the default setup of `electron-builder`.
Is there a way to get this integration working? 

You can see the code here: https://github.com/ipfs/station/tree/better-build

**Comment by dignifiedquire** [comment]
Why was this closed? I never got any response and the problems I had are still standing.

**Comment by unindented** [comment]
@develar can you provide some guidance on what would be needed to get this working?

**Comment by develar** [comment]
@unindented I haven't yet investigate it. If you can provide error message — I can help you to fix it.

**Comment by arekkas** [comment]
Is there currently no way of using electron-builder together with tools such as electron-compile? Other issues mention that this is out of scope, but I'd still like to get this working.

**Comment by develar** [comment]
@arekkas Just compile it using babel — http://electron-builder.slackarchive.io/general/I%252520just%252520discovered%252520that%252520I/

**Comment by arekkas** [comment]
@develar ok so your idea is to run babel on `src`, and change `"main": "src/main/index.js"` to e.g. ` "main": "build/main/index.js",` in the package json?

ps: seems like electron-compile passes the ball to electron-builder, see https://github.com/electron/electron-compile/issues/99

**Comment by develar** [comment]
@arekkas Yes, as was reported by @pronebird.

> seems like electron-compile passes the ball to electron-builder

We are going to support precompilation on our side soon.

**Comment by pronebird** [comment]
@arekkas not without limitations but we do a lot of awesome stuff with babel-cli alone in https://github.com/jschr/electron-react-redux-boilerplate/blob/master/package.json

**Comment by arekkas** [comment]
I see, thanks for pointing me at it. Thank you all for your work on this. At some point, it would probably be cool to have an overview of what to use when for what and how. For a newcomer to the electron ecosystem I have to say it's a steep learning curve :D

**Comment by develar** [comment]
Good news — electron-complie support mostly implemented and will be released in two days. Sorry that it was not implemented so long.

**Comment by develar** [comment]
@pronebird Am I right that ash shim is not required — https://github.com/electron/electron-compile/issues/207 ?

**Comment by pronebird** [comment]
@develar it could be I did something wrong but I had exactly the same problem where this shim was used in production build. I switched to babel since then and never looked back.

**Comment by develar** [comment]
@pronebird Ok. I am waiting @paulcbetts response (https://github.com/electron/electron-compile/issues/207) but it seems shim is not required and electron-builder will not produce shim.

BTW, is ability to provide own transformer for files will be useful for you? I mean — transpile file for production on the fly (electron-builder writes result of transpilation directly to asar file without any intermediate/temp files).

**Comment by pronebird** [comment]
@develar either way I transpile files manually for development so I don't really see the benefit of delegating this to electron-builder.

**Comment by develar** [comment]
Supported in the 16.0.0 — if electron-compile in your dependencies, all files will be transpiled on the fly. Currently, only supported if asar enabled (feel free to file issue to support if asar is disabled).

**Comment by arekkas** [comment]
Awesome!

**Comment by samjbmason** [comment]
It would be good to know what I need to get this new feature to work. I have an app that is using `electron-compile` & `electron-compilers` and building using `electron-builder` with the follwoing options
```js
"scripts": {
    "build": "build"
  },
"build": {
    "productName": "NTS mini",
    "mac": null,
    "asar": true
  },
```

when I build the artifacts are build and i get no errors but when I attempt to open up the app.dmg, I get the following error

```shell
Uncaught Exception:
Error: Cannot find module 'electron-compile'
```

**Comment by develar** [comment]
@samjbmason Is it possible to provide your project for test?

**Comment by samjbmason** [comment]
@develar I have added you as a collaborator if that's ok?

**Comment by pronebird** [comment]
@samjbmason electron apps are generally over 100M due to amount of libraries used in them. They say there is no way to decrease the size.

**Comment by samjbmason** [comment]
@pronebird Ah ok that's not a problem I just wanted to make sure and that it wasn't a symptom of the errors that I was receiving when trying to run the app

**Comment by develar** [comment]
@samjbmason Some errors discovered on your app, some on electron-builder side. I will post details later.

Yes — electron-builder is not going to use easy way because our goal to help users, not just "well, we build something".

**Comment by develar** [comment]
@samjbmason 
* menubar, electron-devtools-installer are used in the app code, but listed in the devDependencies
* electron-compile is used in the app code, but listed in the devDependencies
* consider to not use shim and use electron-compile-preload for your dev mode (and in production it is not required at all — although I see that you use live reload).
* https://github.com/electron/electron/issues/6083
* didn't manage to run app and fix all issue in the end.

Error on electron-builder side fixed, please try 16.0.1

**Comment by samjbmason** [comment]
@develar Thanks very much for that!
Re your points above:
- Good point on menubar I'll move that across, if I conditionally load `electron-devtools-installer` that should sort the issue with requireing it right?
- I moved `electron-compile` into devDependencies because `electron-builder` said I shoud move it to there.
- what is `electron-compile-preload`?

**Comment by develar** [comment]
You should use babel-plugin-transform-inline-imports-commonjs to transform imports to a lazy require statements. It will allow to use imports for any dep, but of course, usage should be only after check that app in the dev mode. Currently, you will get warn about electron-compile dep, but I will fix it later.

Yes, electron-builder asks to move electron-compile to dev deps — it is correct.

About shim — https://github.com/electron-userland/electron-prebuilt-compile

**Comment by develar** [comment]
Reopened as we have the answer — https://github.com/electron/electron-compile/issues/207#issuecomment-288787604

**Comment by arekkas** [comment]
Any idea what's causing this?

```
> electron .


App threw an error during load
C:\workspace\js\sites-app\node_modules\webpack\lib\WebpackOptionsApply.js:209
                        comment = legacy && modern ? "\n/*\n//@ sourceMappingURL=[url]\n//# sourceMappingURL=node_modules\webpack\lib\[url]\n*\" :
                                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: Invalid or unexpected token
```

**Comment by develar** [comment]
@arekkas electron-buidler version?

**Comment by arekkas** [comment]
@develar 16.3.0 - also I'm using `electron-prebuilt-compile`. I'm currently trying to remove all the fragments from my earlier toolchain which included babel-cli etc

**Comment by develar** [comment]
@arekkas If you can, please provide project to test (you can send me link privately using slack).

**Comment by arekkas** [comment]
@develar thanks for this, I'll try a couple more things and if I'm unable to figure it out I'll come back to that :) Just wanted to check if there's any additional wiring required for electron-compiler, but I guess this should work out of the box right?

**Comment by develar** [comment]
@arekkas  Is file on disk is equals to extracted file `node_modules\webpack\lib\WebpackOptionsApply.js`? Install asar globally (`npm install asar -g`) and check `asar e resources/app.asar _dir to unpack_` where `resources/app.asar` is path to your app.asar

> but I guess this should work out of the box right?

Yes.

**Comment by arekkas** [comment]
~I'm not sure if I understood you correctly, but yes, that is the right path. My project is kinda special, because I use webpack within the project itself to compile third party code using said webpack, which is why webpack, css-loader, babel are first class citizen in my `package.json` that are actually required by some of my application code. Imagine it as a tool that compiles your es6 project using electron and some ui stuff.~

By the way, I'm not talking about the bundle here, but simply `electron .`. Actually this might be an issue caused by electron-compile, not electron-builder, as electron-builder is not in the toolchain 🙄 . I'll dive into this more and come back when I have something concrete - sorry for wasting your time :)

**Comment by arekkas** [comment]
Is there a way to *disable* electron-compile?

**Comment by develar** [comment]
@arekkas Will be added today. Why do you want to disable?

**Comment by develar** [comment]
`electronCompile` option added.

> Whether to use [electron-compile](http://github.com/electron/electron-compile) to compile app. Defaults to `true` if `electron-compile` in the dependencies. And `false` if in the `devDependencies` or doesn't specified.

Will be released soon.

**Comment by arekkas** [comment]
Because for now I'm sticking with my babel flow, but I still need electron-compile for my e2e tests to work, so it's in the devDependencies which causes the package to crash with "electron-compile not found"

**Comment by arekkas** [comment]
Nice, thanks for the quick update!
